assume/markets/clearing_algorithms/complex_clearing.py

```diff
@@ -0,0 +1,147 @@
+import logging
+from itertools import groupby
+from operator import itemgetter
+
+from assume.common.market_objects import MarketProduct, Orderbook
+from assume.markets.base_market import MarketRole
+
+log = logging.getLogger(__name__)
+
+SOLVERS = ["glpk", "cbc", "gurobi", "cplex"]
+
+
+def pay_as_clear_opt(
+    market_agent: MarketRole,
+    market_products: list[MarketProduct],
+):
+    import pyomo.environ as pyo
+    from pyomo.opt import SolverFactory, check_available_solvers
+
+    accepted_orders: Orderbook = []
+    rejected_orders: Orderbook = []
+    meta = []
+
+    market_getter = itemgetter("start_time", "end_time", "only_hours")
+    market_agent.all_orders.sort(key=market_getter)
+
+    if len(market_agent.all_orders) == 0:
+        return accepted_orders, meta
+
+    model = pyo.ConcreteModel()
+    model.dual = pyo.Suffix(direction=pyo.Suffix.IMPORT_EXPORT)
+
+    model.T = pyo.Set(
+        initialize=[market_product[0] for market_product in market_products],
+        doc="timesteps",
+    )
+    model.K = pyo.Set(
+        initialize=[order["bid_id"] for order in market_agent.all_orders], doc="bids"
+    )
+
+    model.P = pyo.Var(model.K, domain=pyo.Reals)
+
+    model.energy_balance = pyo.ConstraintList()
+    model.min_max_power = pyo.ConstraintList()
+
+    for product, product_orders in groupby(market_agent.all_orders, market_getter):
+        if product not in market_products:
+            rejected_orders.extend(product_orders)
+            # log.debug(f'found unwanted bids for {product} should be {market_products}')
+            continue
+
+        product_orders = list(product_orders)
+
+        expr = sum(model.P[order["bid_id"]] for order in product_orders) == 0
+        model.energy_balance.add(expr)
+
+        for order in product_orders:
+            if "acceptance_ratio" in market_agent.marketconfig.additional_fields:
+                if order["volume"] > 0:
+                    min_max_power_expr = (
+                        order["acceptance_ratio"] * order["volume"],
+                        model.P[order["bid_id"]],
+                        order["volume"],
+                    )
+                else:
+                    min_max_power_expr = (
+                        order["volume"],
+                        model.P[order["bid_id"]],
+                        order["acceptance_ratio"] * order["volume"],
+                    )
+
```

Comment:
doesn't this miss out the option to not use a bid at all?
As I see, the value should always be between `acceptance_ratio * volume` and `volume`, but it should also be possible to set the volume to 0?

Reply:
yes true, missed that. the nodal clearing one has the same problem in this case

Reply:
I think it should work as 0 is respected here:

https://github.com/assume-framework/assume/blob/0b600abce5d5671be6596a8e5e233d5761b0f755/assume/markets/clearing_algorithms/pyomo_clearing.py#L147-L150

But adding an additional test for this would also be good, as I did not test this yet